### ISSUE TITLE
default value support

### ISSUE THREAD

**Issue by Roger-luo** [JuliaData/StructTypes.jl#53]
Hi, I'm wondering how can I define a field has a corresponding default value, I can't find a trait that defines this property, thus as a result JSON3 will treat field not specified with a value in the payload as `Nothing` instead of using the default value and fail to create the struct type. I use this type of convention quite frequently in structs created by [Configurations.jl](https://github.com/Roger-luo/Configurations.jl), I'm thinking about auto generating the overloads for StructTypes in Configurations. Ideally I'm thinking if the `construct` method will be able to use the keyword constructor created by things like `Configurations.@option` or `Base.@kwdef` so that the default value will be handled automatically

An example of this type of struct looks like the following (part of this is used in Pluto)

```julia
@option mutable struct CompilerOptions
    compile::Union{Nothing,String} = nothing
    sysimage::Union{Nothing,String} = nothing
    banner::Union{Nothing,String} = nothing
    optimize::Union{Nothing,Int} = nothing
    math_mode::Union{Nothing,String} = nothing

    # notebook specified options
    # the followings are different from
    # the default julia compiler options

    # we use nothing to represent "@v#.#"
    project::Union{Nothing,String} = "@."
    # we don't load startup file in notebook
    startup_file::Union{Nothing,String} = "no"
    # we don't load history file in notebook
    history_file::Union{Nothing,String} = "no"

    threads::Union{Nothing,String,Int} = get_nthreads()
end

@option struct DistributedOptions
    tunnel::Bool = false
    multiplex::Bool = false
    ssh::Maybe{String} = nothing
    sshflags::Maybe{String} = nothing
    max_parallel::Int = 10
    shell::Symbol = :posix
    dir::String = pwd()
    enable_thread_blas::Bool = false
    exename::String = joinpath(Sys.BINDIR, "julia")
    topology::Symbol = :all_to_all
    lazy::Bool = true
    env::Maybe{Vector{Pair{String, String}}} = nothing
end

@option struct WorkspaceInstance
    name::Symbol = gensym(:workspace)
    uuid::UUID = uuid1()
    compiler::CompilerOptions = CompilerOptions()
    distributed::DistributedOptions = DistributedOptions()
    # scripts that is evaluated in this workspace instance
    linked_scripts::Vector{String} = String[]
end
```

**Comment by quinnj** [comment]
Yeah, we've talked about adding a new StructType like `KeywordArgs` that would support `Base.@kwdef`-built structs.

**Comment by onetonfoot** [comment]
This macro seems todo the job

```julia
@option struct A 
    x::String = "hi"
    y::Vector{String} = []
end

namedtuple(d::AbstractDict) = (;Dict(Symbol(k) => namedtuple(v) for (k, v) in d)...)
namedtuple(x::Array) = namedtuple.(x)
namedtuple(x) = x

macro json3(T) 
    quote
        StructTypes.StructType(::Type{T}) where T <: $T = JSON3.CustomStruct()
    function StructTypes.lower(x::T) where T <: $T
            Dict(
            k => getproperty(x, k) for k in fieldnames($T)
        )
        end
    StructTypes.lowertype(::Type{T}) where T <: $T = Dict{Symbol,Any}
    StructTypes.construct(::Type{T}, x) where T <: $T = $T(;namedtuple(x)...)
    end |> esc
end

@json3 A

JSON3.read("""{"y" : ["a"]}""", A)

```

**Comment by quinnj** [comment]
Very cool @onetonfoot! Thanks for sharing!

**Comment by Roger-luo** [comment]
thanks! Just want to provide a more complete code snippet based on @onetonfoot , this would use `from_dict` constructor defined in Configurations to convert a dictionary-like type to `T`

```julia
using StructTypes
using Configurations
StructTypes.StructType(::Type{<:MyOptionType}) = StructTypes.CustomStruct()
StructTypes.lower(x::MyOptionType) = to_dict(x, JSONStyle)
StructTypes.lowertype(::Type{<:MyOptionType}) = OrderedDict{String, Any}
StructTypes.construct(::Type{T}, x) where {T <: MyOptionType} = from_dict(T, x)
```

but I'm guessing this would lost some of the composibility by converting `Object` to an `OrderedDict{String, Any}` when parsing to `MyOptionType` using `JSON3` which might have some impact on performance? I'm not sure

**Comment by heyx3** [comment]
If your fields need to be deserialized by `StructTypes` themselves, then I don't think the macro given by @onetonfoot would work. However, I was able to overload an internal version of `StructTypes.construct()` to get this working. You could probably turn it into a macro or `@generated` with a little work.

Here's the implementation for my struct, `Base.@kwdef struct Sampler`, which has 8 fields:

    function StructTypes.construct(values::Vector{Any}, T::Type{<:Sampler})
        kw_args = NamedTuple()
        if isassigned(values, 1)
            kw_args = (kw_args..., wrapping=values[1])
        end
        if isassigned(values, 2)
            kw_args = (kw_args..., pixel_filter=values[2])
        end
        if isassigned(values, 3)
            kw_args = (kw_args..., mip_filter=values[3])
        end
        if isassigned(values, 4)
            kw_args = (kw_args..., anisotropy=values[4])
        end
        if isassigned(values, 5)
            kw_args = (kw_args..., mip_offset=values[5])
        end
        if isassigned(values, 6)
            kw_args = (kw_args..., mip_range=values[6])
        end
        if isassigned(values, 7)
            kw_args = (kw_args..., depth_comparison_mode=values[7])
        end
        if isassigned(values, 8)
            kw_args = (kw_args..., cubemap_seamless=values[8])
        end

        return T(; kw_args...)
    end

**Comment by quinnj** [comment]
Implemented in #79
